### PR TITLE
Free gifted device allocations even when wrapping host memory

### PR DESCRIPTION
**Pull request by abadams**
This builds on PR #2036 . The diff will be much smaller once that is merged.

In situations like this:

        {
            int scratch[80];
            Buffer<int> output_wrapped(scratch, 80);
            some_pipeline(output_wrapped);
        }

We should free any device allocation that the pipeline attaches to the Buffer, even though we're not responsible for freeing host memory.

Note that gpu_object_lifetime wasn't testing anything before, as it was linking to a standalone runtime without -debug on, and it required -debug on to do any actual checking. This PR fixes that too, and uses it to test the change to device memory ownership.

### CODE REVIEW

**Comment by abadams** [comment]
Now that the other PR is in, this is reviewable.

File: src/runtime/HalideBuffer.h [hunk 1]
```diff
@@ -674,22 +675,19 @@ class Buffer {
...
     /** Drop reference to any owned device memory, possibly freeing it
-     * if this buffer held the last reference to it. Does nothing if
-     * this buffer did not allocate its own memory. Asserts that
+     * if this buffer held the last reference to it. Asserts that
      * device_dirty is false. */
     void device_deallocate() {
-        if (manages_memory()) {
-            decref_dev();
-        }
+        decref_dev();
```
Comment (jingpu) [inline src/runtime/HalideBuffer.h]: Is this the only functional change? Why does it fix the issue?

Comment (jingpu) [inline src/runtime/HalideBuffer.h]: Nevermind. I found the change in decref().

File: src/runtime/HalideBuffer.h [hunk 2]
```diff
@@ -150,16 +150,17 @@ class Buffer {
...
 
 private:
     /** Increment the reference count of any owned allocation */
     void incref() const {
-        if (!manages_memory()) return;
-        alloc->ref_count++;
+        if (owns_host_memory()) {
+            alloc->ref_count++;
+        }
         if (buf.device) {
```
Comment (jingpu) [inline src/runtime/HalideBuffer.h]: For the example in the PR description, will this code path be exercised before `output_wrapped` is destructed? There is no copy or move ctor of Buffer called after the Halide pipeline allocates a device buffer internally. Maybe I missed something.

Comment (abadams) [inline src/runtime/HalideBuffer.h]: It won't be, but decref_dev will now be called unconditionally by decref, which will notice that there's a non-zero device field and free it.

Comment (jingpu) [inline src/runtime/HalideBuffer.h]: It makes sense. Thanks!

**Pull request merged**